Automattic/wp-calypso
# Editor: Reset edits when navigating away from new post without saving

This pull request seeks to resolve an issue where navigating away from a new draft to My Sites will cause the "Add" buttons adjacent to custom post types to link incorrectly to the previously edited post.

__Implementation notes:__

The issue occurs in [the `getEditorPath` selector](https://github.com/Automattic/wp-calypso/blob/2107ee459ad78d8d9e5427c7b65eb0f8b3082a82/client/state/ui/editor/selectors.js#L52), and is a result of how we treat `getEditedPost` as meaning the unsaved draft [when a null post ID is passed](https://github.com/Automattic/wp-calypso/blob/2107ee459ad78d8d9e5427c7b65eb0f8b3082a82/client/state/posts/selectors.js#L206), which is further complicated by our logic for generating an editor editor path for a new custom post type by [using a null post ID](https://github.com/Automattic/wp-calypso/blob/2107ee459ad78d8d9e5427c7b65eb0f8b3082a82/client/my-sites/sidebar/publish-menu.jsx#L205). We should probably seek to create a better distinction between these use-cases, by either creating a standalone `getEditorNewPostPath` or not considering edits in generating the editor path (changed previously in 16976b38fe34ce1d3618213214b4d26631d7d375).

Regardless of improvements around `getEditorPath`, it's reasonable to expect that we should be resetting unsaved draft edits when navigating away from the editor, or at the very least when navigating to a new editor draft session. The changes here add a new exit route hook, discarding post edits if the post was not saved.

Note that the only post edit state currently tracked in Redux state is the post type, and that this issue does not currently affect production environments since the `getEditorPath` selector is not used in any environment aside from development.

__Testing instructions:__

Verify that the correct "Add" link is shown for a custom post type after navigating away from the post editor.

1. Navigate to the [post editor](http://calypso.localhost:3000/post)
2. Select a site where a custom post type is enabled
3. Click My Sites in the master bar
4. Click the add button next to the custom post type sidebar item
5. Note that you are navigated to a new editor draft for that custom post type
